### DISCUSSION TITLE
Loading Screen and Quitting game?

### DISCUSSION THREAD

**Discussion by dharry1968** [jeffreylanters/react-unity-webgl#137] [Question and Answer] [answered]
Would you mind helping me with 2 questions?
 
1.	When my game loads, there is no loading screen, it’s just a grey window and then after a while the game starts. Do you know what might cause that?

2.	When I exit the game by calling Application.Quit(), the game quits as it should but just sits on a blank page… do you know how to have it go back to my site home page or something?

Regards,
Dean

**Comment by Rickleung** [comment]
1. did you use the default unity logo from the player setting inside unity.?
2. I think you can create a button and pass a function from c # to javascript and tell it to load a certain web page.

P.S. Would you mind to share your template....? I having problems when trying to implement react into my webgl build project.....I am confused whethere I should put the

 import React from "react";
import Unity, { UnityContext } from "react-unity-webgl";

inside my index.html or I should create a new js file?? Thanks....

Rick

**Reply by dharry1968** [reply]
Hi Rick,

1. Yes, I havn't changed any files or loading screens at all... if I call it outside of React it shows the loading page, just not from inside React.
2. That's a good idea, didn't think of that :) i'll give that a go...

I have a index.js in the game branch folder that contains this:

import React from "react";
import Unity, { UnityContent } from "react-unity-webgl";

export default function App() {
    let unityContent = new UnityContent(
        "../static/unity/Build/Builds.json",
        "../static/unity/Build/UnityLoader.js"
    );
    return (
        <div><Unity unityContent={unityContent} /></div>
    );
}

Cheers
Dean

**Comment by jeffreylanters** [comment, marked as answer]
Hi Dean! Thanks for opening this discussion. I'm afraid you're misunderstanding the purpose of this module and the use of React.

The React Unity WebGL module is meant to implement your Unity Build into your React Application so you can create your own interface and everything around it. I think you're confusing it with the Templating-Tool Unity has added to their Build-Pipeline. If you want a pre-game loading screen, you have to build one yourself using React, as the loading screen you're referring to is not included in your Unity Build, but is living inside of the HTML which Unity's Templating-Tool has created.

It's simply not the purpose of a module like this to serve things such as a loading screen, but instead to serve you - the developer - the tools needed to create one yourself. For more information about building a loading screen, please see [this section in the Wiki](https://github.com/elraccoone/react-unity-webgl/wiki/Communication-Unity-on-Progress) for further details. A basic example would look something like:

```js
this.unityContext.on("progress", progression => {
  this.setState({ percent: progression * 100 });
});
...
<div>
  <div>{`Loading... ${percent}%...`}</div>
  <Unity unityContext={this.unityContext} />
</div>
```

The same goes for your second point, when you have quitted your application, it is  up to you - the developer - to act on this. All Quiting the Application does is freeing up memory for your browser. You can add behaviour by adding an event listener to the Quit event. Please [see this section in the Wiki](https://github.com/elraccoone/react-unity-webgl/wiki/Communication-Unity-on-Quitted) for further details. A basic example would look something like:

```js
this.unityContext.on("quitted", () => {
  window.location.href = "...";
});
```

I hope this clears up some of your questions. Best of luck with your project, and stay safe! If you're having any other questions feel free to reply on this thread.

**Reply by dharry1968** [reply]
Thanks Jeffrey, slowly starting to understand... React is very new to me, sorry...  whilst I get what your saying above, all the examples I see are using UnityContext... I am using UnityContent instead as that was the only thing that seemed to work, trying the UnityContext way just gives me an error. 

Is it possible to use these events this way somehow as well?

**Reply by jeffreylanters** [reply]
No need to say you're sorry, I'm really happy a fellow game developer is getting interested in web development. I'm here to help. About the UnityContext and UnityContent, which version of the Module and which version of Unity are you using? Both of these events are the same when using the older UnityContent.

**Reply by dharry1968** [reply]
I'm using Unity 2019.4.16 and @7.x of the module...

**Reply by jeffreylanters** [reply]
Thanks Dean, I see. This is a valid combination, you can use most of the same events in the UnityContent as you can in the UnityContext. If you'd like to use the latest available methods, you'd have to upgrade to Unity 2020 (Which is recommended due to their improved WebGL build support). Let me know if you're able to implement the loading screen and application quitting. Feel free to ask help!

**Reply by jeffreylanters** [reply]
If you're not able to implement these features, I can always help you with your project personally. I'm happy to help a fellow game developer get into web development. 😄

**Reply by dharry1968** [reply]
Hi Jeffrey, it's coming along... I think i'm a little confused somewhere though... I have this code which is supposed to show a loading spinner, I do get the "Loading Progress..." on the console so I know it's getting there, it just doesn't show the spinner... also, I read somewhere that in order to get the "quitted" to fire you need to call an event from Unity just before Application.Exit(), is that as simple as just SendMessage("quitted")?

        this.unityContent = new UnityContent(
            "../static/unity/Build/Builds.json",
            "../static/unity/Build/UnityLoader.js"
        );

        this.unityContent.on("quitted", () => {
            console.log("Quitting Now");
            window.location.href = "https://darkgalaxies.io";
        });
        
        this.unityContent.on("progress", progression => {
            console.log("Loading Progress...");
            <div style={{ position: 'relative', width: 200, height: 200 }}>
                <Loading animate full />
            </div>
        });

    return (
        <div>
            <div><Unity unityContent={unityContent} /></div>
        </div>
    );

**Reply by jeffreylanters** [reply]
Hi Dean! 😄

You're getting on the right track. Looking at your attempt, I do recommend following a basic React course. This might help you in getting used on how to use the React state and how JSX works in general. I'll rewrite your code to make it work below.

It is not possible to render HTML (JSX) in a function like that. What I'm going to do is create a state, which contains the progress and a boolean to define whether loading is done or not. I check this state in my render function in order to act on it. I recommend watching [this video](https://www.youtube.com/watch?v=4ORZ1GmjaMc) to learn more about the state within react.

```js
    this.state = {
        progression: 0,
        isLoaded: false
    };

    this.unityContent = new UnityContent(
        "../static/unity/Build/Builds.json",
        "../static/unity/Build/UnityLoader.js"
    );
    
    this.unityContent.on("progress", progression => {
        console.log("Loading Progress...");
        this.setState({ progression: progression });
    });

    this.unityContent.on("loaded", progression => {
        this.setState({ isLoaded: true });
    });

return (
    <div>
        <Unity unityContent={unityContent} />
       {this.state.isLoaded == false && <div>
          <div style={{ position: 'relative', width: 200, height: 200 }}>
              <Loading animate full />
          </div>
       </div>}
    </div>
);
```

**Reply by jeffreylanters** [reply]
And about the on Quitted event listener, this event listener will only invoke when you unmount the Unity component (ie let's say when you navigate to another page), or when you quit the application from within the Application.Quit method will be invoked.

Best of luck!

**Reply by dharry1968** [reply]
Thanks Jeffrey, i'll deffintley jump on a React course to get a better understanding of it... just tried to run that inj quickly but i'm getting some issue about a theme, i'll take another look tomorrow. 

I am calling Application.Quit from Unity but it's not firing, i'll take another look at that tomorrow as well :)

**Reply by dharry1968** [reply]
Hi Jeffrey, ok, I have been on and off this all day but still can't get it to show the loading spinner... at the moment I have the following:

export class App  extends React.Component {
    constructor() {
        this.state = {
            progression: 0,
            isloaded: false
        };

        this.unityContent = new UnityContent(
            "../static/unity/Build/Builds.json",
            "../static/unity/Build/UnityLoader.js"
        );

        this.unityContent.on("quitted", () => {
            console.log("Quitting Now");
            window.location.href = "https://darkgalaxies.io";
        });

        this.unityContent.on("progress", progression => {
            console.log("Loading Progress...");
            this.setState({ progression: progression });
        });

        this.unityContent.on("loaded", progression => {
            this.setState({ isloaded: true });
        });

        return (
            <div>
                <div><Unity unityContent={unityContent} /></div>
                {this.state.isloaded == false && <div>
                    <div style={{ position: 'relative', width: 200, height: 200 }}>
                        <Loading animate full />
                    </div>
                </div>}
            </div>
        );
    }
}

but I keep on getting the error: Cannot read property 'getInitialProps' of undefined 

I think I understand that this is coming from Next.js but not quite sure why it is telling me that. What have I missed?

**Reply by jeffreylanters** [reply]
Hi Dean! I have never used NextJS before. I'm not quitte sure where this error is coming from. Can you try figuring out which component or module throws this error? Maybe by looking at the stack trace, or otherwise slowly stripping your component down. Thanks!

**Reply by jeffreylanters** [reply]
Hi Dean! How are you doing? Did you manage to get your project working, or need some more help? -- I wish you a Happy new-year!

**Reply by dharry1968** [reply]
Hi Jeffrey, happy new year!! 

Actually, no I havn't as yet... I did try and got stuck and then was caught up in other things. I would love some help if it's on offer! That would be fantastic. 

There are 3 things I am stuck on...

1. The loading screen
2. Redirect after quit
3. Passing in parameters from React at game start

Regards,
Dean
